miniapps/shifted_methods/dist_solver.cpp

```diff
@@ -0,0 +1,784 @@
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "dist_solver.hpp"
+
+using namespace mfem;
+
+void DiffuseField(ParGridFunction &field, int smooth_steps)
+{
+   // Setup the Laplacian operator.
+   ParBilinearForm *Lap = new ParBilinearForm(field.ParFESpace());
+   Lap->AddDomainIntegrator(new DiffusionIntegrator());
+   Lap->Assemble();
+   Lap->Finalize();
+   HypreParMatrix *A = Lap->ParallelAssemble();
+
+   HypreSmoother *S = new HypreSmoother(*A,0,smooth_steps);
+   S->iterative_mode = true;
+
+   Vector tmp(A->Width());
+   field.SetTrueVector();
+   Vector fieldtrue = field.GetTrueVector();
+   tmp = 0.0;
+   S->Mult(tmp, fieldtrue);
+
+   field.SetFromTrueDofs(fieldtrue);
+
+   delete A;
+   delete S;
+   delete Lap;
+}
+
+void DistanceSolver::ScalarDistToVector(ParGridFunction &dist_s,
+                                        ParGridFunction &dist_v)
+{
+   ParFiniteElementSpace &pfes = *dist_s.ParFESpace();
+   MFEM_VERIFY(pfes.GetOrdering()==Ordering::byNODES,
+               "Only Ordering::byNODES is implemented.");
+
+   const int dim = pfes.GetMesh()->Dimension();
+   const int size = dist_s.Size();
+
+   ParGridFunction der(&pfes);
+   Vector magn(size);
+   magn = 0.0;
+   for (int d = 0; d < dim; d++)
+   {
+      dist_s.GetDerivative(1, d, der);
+      for (int i = 0; i < size; i++)
+      {
+         magn(i) += der(i) * der(i);
+         // The vector must point towards the level zero set.
+         dist_v(i + d*size) = (dist_s(i) > 0.0) ? -der(i) : der(i);
+      }
+   }
+
+   const double eps = 1e-16;
+   for (int i = 0; i < size; i++)
+   {
+      const double vec_magn = std::sqrt(magn(i) + eps);
+      for (int d = 0; d < dim; d++)
+      {
+         dist_v(i + d*size) *= fabs(dist_s(i)) / vec_magn;
+      }
+   }
+}
+
+void DistanceSolver::ComputeVectorDistance(Coefficient &zero_level_set,
+                                           ParGridFunction &distance)
+{
+   ParFiniteElementSpace &pfes = *distance.ParFESpace();
+   MFEM_VERIFY(pfes.GetVDim() == pfes.GetMesh()->Dimension(),
+               "This function expects a vector ParGridFunction!");
+
+   ParFiniteElementSpace pfes_s(pfes.GetParMesh(), pfes.FEColl());
+   ParGridFunction dist_s(&pfes_s);
+   ComputeScalarDistance(zero_level_set, dist_s);
+   ScalarDistToVector(dist_s, distance);
+}
+
+void HeatDistanceSolver::ComputeScalarDistance(Coefficient &zero_level_set,
+                                               ParGridFunction &distance)
+{
+   ParFiniteElementSpace &pfes = *distance.ParFESpace();
+
+   auto check_h1 = dynamic_cast<const H1_FECollection *>(pfes.FEColl());
+   MFEM_VERIFY(check_h1 && pfes.GetVDim() == 1,
+               "This solver supports only scalar H1 spaces.");
+
+   // Compute average mesh size (assumes similar cells).
+   double dx, loc_area = 0.0;
```

Comment:
I get a compiler warning that `dx` is uninitialized in the `default` case of the switch statement below.